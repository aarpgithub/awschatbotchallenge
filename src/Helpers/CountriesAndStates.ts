import states = require("us-state-codes");

export const stateNameToCode = (state: string) => {
  let stateCode = states.sanitizeStateName(state);
  if (stateCode == null) {
    throw Error(`Error missing field state.  Input: ${state}`);
  }
  stateCode = states.getStateCodeByStateName(stateCode);
  if (stateCode == null) {
    throw Error("Error converting state to code.  Input: ${state}");
  }
  return stateCode;
};
