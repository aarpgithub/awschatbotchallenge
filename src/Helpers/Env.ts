export class Env {
  static getString(key: string): string {
    return process.env[key] || "";
  }

  static getInt(key: string): Number {
    return parseInt(process.env[key] || "0");
  }
}
