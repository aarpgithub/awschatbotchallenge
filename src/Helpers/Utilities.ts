import { isNil } from "lodash";
import { createHash } from "crypto";
import traverse = require("traverse");
import { union as lodash_union, flatten as lodash_flatten } from "lodash";

/**
 * Remove any keys with undefined or null values from obj and return obj.  
 * !! Mutates argument !!
 * 
 * example: 
 *  const obj = {a: null, b: 1}
 *  const obj2 = excludeNil(obj)
 *  console.log(obj2)
 * -> "{ b: 1}"
 * @param obj 
 */
export function excludeNil<T>(obj: T): T {
  for (const key in obj) {
    if (isNil(obj[key])) {
      delete obj[key];
    }
  }
  return obj;
}

export const sha256 = (str: string): string => {
  const hash = createHash("sha256");
  hash.update(str);
  return hash.digest("hex");
};

export const sha256checksum = (obj: Object): string => {
  return sha256(JSON.stringify(obj));
};

export const removeKeysRecursive = (obj: any, keys: Array<String>) => {
  traverse(obj).forEach(function(e: any) {
    if (this.key && containsKey(this.key, keys)) {
      this.remove(true);
    }
  });
};

export const containsKey = (key: string, keys: Array<String>) => {
  if (keys.indexOf(key) > -1) {
    return true;
  }
  return false;
};

export const snooze = (ms: number) => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

export const union = lodash_union;
export const flatten = lodash_flatten;
