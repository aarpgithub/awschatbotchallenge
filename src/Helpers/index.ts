export * from "./CountriesAndStates";
export * from "./Env";
export * from "./Utilities";
