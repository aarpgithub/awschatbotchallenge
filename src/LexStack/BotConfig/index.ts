export * from "./LambdaBotConfig";
export * from "./ServerlessBotConfig";
