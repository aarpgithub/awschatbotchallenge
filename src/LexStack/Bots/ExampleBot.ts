import {
  SlotTypeDefinition,
  IntentDefinition,
  BotDefinition
} from "../LexService";
import { ExampleBotSlotType } from "../SlotTypes";
import { ExampleBotIntent } from "../Intents";

export class ExampleBot extends BotDefinition {
  _name = "ExampleBot";

  description = "A bot used for testing purposes";

  get intents() {
    return [new ExampleBotIntent(this.c)];
  }

  clarificationPrompt = {
    messages: [
      {
        contentType: "PlainText",
        content:
          "I'm sorry I didn't understand that. Can you please ask it in a different way?"
      }
    ],
    maxAttempts: 2
  };

  abortStatement = {
    messages: [
      {
        contentType: "PlainText",
        content: "Sorry, I could not understand. Goodbye."
      }
    ]
  };

  idleSessionTTLInSeconds = 120;
  voiceId = "Salli";

  get aliases(): Array<string> {
    return [];
  }
}
