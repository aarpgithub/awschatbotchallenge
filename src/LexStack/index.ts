export * from "./LexService";
export * from "./LexMessages";
export * from "./BotConfig";
export * from "./Bots";
export * from "./Intents";
export * from "./SlotTypes";
