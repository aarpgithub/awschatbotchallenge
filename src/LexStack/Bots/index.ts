export * from "./ExampleBot";
export * from "./HealthCareBot";

import { BotConfiguration, BotDefinition } from "../LexService";
import { ExampleBot, HealthCareBot } from "../Bots";

// silly interface for removing some type information ...
interface BotMapContainer {
  [k: string]: BotDefinition;
}

export const BotMap = (c: BotConfiguration) => {
  return {
    ExampleBot: new ExampleBot(c),
    HealthCareBot: new HealthCareBot(c)
  } as BotMapContainer;
};

export const AllBots = (c: BotConfiguration) => {
  const bots = Object.values(BotMap(c));
  if (bots == null) {
    throw Error("Error force unwrapping optional");
  }
  return bots;
};
