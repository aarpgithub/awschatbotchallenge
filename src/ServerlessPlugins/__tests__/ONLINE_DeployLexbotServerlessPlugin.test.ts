import { LexService, ServerlessBotConfig } from "../../LexStack";
import {
  ServerlessPatched,
  DeploymentStage
} from "../../ServerlessPlugins/JestServerlessConfig";

import { ExampleBot } from "../../LexStack/Bots";
import { removeKeysRecursive } from "../../Helpers";

let serverlessConfig: ServerlessPatched | null = null;

beforeAll(async () => {
  serverlessConfig = new ServerlessPatched("stage");
  await serverlessConfig.init();
  await serverlessConfig.run();
});

it("Expected BotConfig should be pulled correctly from the serverless config", () => {
  const botConfig = new ServerlessBotConfig(serverlessConfig, {}, true);
  const { stage, region, ownerId, profile, lexTestMode } = botConfig;
  expect({ stage, region, ownerId, profile, lexTestMode }).toMatchSnapshot();
});

describe("ONLINE:", () => {
  it("Deployed Bot definition should match expectations", async () => {
    const botConfig = new ServerlessBotConfig(serverlessConfig, {}, true);
    const lexService = new LexService(botConfig);
    const bot = new ExampleBot(botConfig);

    await lexService.updateFullBot(bot);
    const deployed: any = {};

    deployed.slotTypeDefinitions = [];
    for (const d of bot.slotTypeDefinitions) {
      deployed.slotTypeDefinitions.push(
        await lexService.getSlotType({
          name: d.name,
          version: "$LATEST"
        })
      );
    }

    deployed.intentDefinitions = [];

    for (const d of bot.intents) {
      deployed.intentDefinitions.push(
        await lexService.getIntent({
          name: d.name,
          version: "$LATEST"
        })
      );
    }

    deployed.botDefinition = await lexService.getBot({
      name: bot.name,
      versionOrAlias: "$LATEST"
    });

    // remove transient properties
    removeKeysRecursive(deployed, [
      "createdDate",
      "lastUpdatedDate",
      "checksum"
    ]);

    expect(deployed).toMatchSnapshot();
  });
});
