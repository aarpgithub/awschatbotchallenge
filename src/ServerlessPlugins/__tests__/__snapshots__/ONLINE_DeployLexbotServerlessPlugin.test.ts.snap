// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expected BotConfig should be pulled correctly from the serverless config 1`] = `
Object {
  "lexTestMode": true,
  "ownerId": "321144413607",
  "profile": "aarpai-middleware-lexbot-stage",
  "region": "us-east-1",
  "stage": "stage",
}
`;

exports[`ONLINE: Deployed Bot definition should match expectations 1`] = `
Object {
  "botDefinition": Object {
    "abortStatement": Object {
      "messages": Array [
        Object {
          "content": "Sorry, I could not understand. Goodbye.",
          "contentType": "PlainText",
        },
      ],
      "responseCard": null,
    },
    "childDirected": false,
    "clarificationPrompt": Object {
      "maxAttempts": 2,
      "messages": Array [
        Object {
          "content": "I'm sorry I didn't understand that. Can you please ask it in a different way?",
          "contentType": "PlainText",
        },
      ],
      "responseCard": null,
    },
    "description": "A bot used for testing purposes",
    "failureReason": null,
    "idleSessionTTLInSeconds": 120,
    "intents": Array [
      Object {
        "intentName": "testmode_ExampleBotIntent",
        "intentVersion": "$LATEST",
      },
    ],
    "locale": "en-US",
    "name": "testmode_ExampleBot",
    "status": "READY",
    "version": "$LATEST",
    "voiceId": "Salli",
  },
  "intentDefinitions": Array [
    Object {
      "description": "This is just a definition for testing",
      "dialogCodeHook": Object {
        "messageVersion": "1.0",
        "uri": "arn:aws:lambda:us-east-1:321144413607:function:HealthCareTool",
      },
      "fulfillmentActivity": Object {
        "codeHook": Object {
          "messageVersion": "1.0",
          "uri": "arn:aws:lambda:us-east-1:321144413607:function:HealthCareTool",
        },
        "type": "CodeHook",
      },
      "name": "testmode_ExampleBotIntent",
      "parentIntentSignature": null,
      "sampleUtterances": Array [
        "I would like to calculate premium increase",
        "What would be my premium",
        "What would be my health care premium",
      ],
      "slots": Array [
        Object {
          "description": "Choose 10 or 100",
          "name": "ExampleBotSlotType",
          "priority": 1,
          "responseCard": null,
          "sampleUtterances": Array [
            "I'll take {ExampleBotSlotType}",
            "My favorite number is {ExampleBotSlotType}",
          ],
          "slotConstraint": "Required",
          "slotType": "AMAZON.NUMBER",
          "slotTypeVersion": null,
          "valueElicitationPrompt": Object {
            "maxAttempts": 2,
            "messages": Array [
              Object {
                "content": "Do you want 10 or 100?",
                "contentType": "PlainText",
              },
            ],
            "responseCard": null,
          },
        },
        Object {
          "description": "Represents a US State",
          "name": "StateNameSlotType",
          "priority": 2,
          "responseCard": null,
          "sampleUtterances": Array [
            "I live in {StateNameSlotType}",
            "I come from {StateNameSlotType}",
          ],
          "slotConstraint": "Required",
          "slotType": "AMAZON.US_STATE",
          "slotTypeVersion": null,
          "valueElicitationPrompt": Object {
            "maxAttempts": 2,
            "messages": Array [
              Object {
                "content": "Provide your state",
                "contentType": "PlainText",
              },
            ],
            "responseCard": null,
          },
        },
      ],
      "version": "$LATEST",
    },
  ],
  "slotTypeDefinitions": Array [
    Object {
      "description": "Choose 10 or 100",
      "enumerationValues": Array [
        Object {
          "value": "100",
        },
        Object {
          "value": "10",
        },
      ],
      "name": "ExampleBotSlotType",
      "version": "$LATEST",
    },
    Object {
      "description": "Represents a US State",
      "name": "StateNameSlotType",
      "version": "$LATEST",
    },
  ],
}
`;
